From 8a3e5e720e37c9883750af68c826baff4b1a887a Mon Sep 17 00:00:00 2001
From: mickhuang <mickhuang@ami.com>
Date: Tue, 12 Aug 2025 14:13:03 +0800
Subject: [PATCH] Revert SGPIO slave to control parallel data for A0 Chip

ASPEED Commit hash
- b388c7e16e395082ec2202b43c0b8cd99d777e31
- 1dc682eeb88bb30f86490c1427fd9faa7c7f6e16
- 333ca7e070726409a06152c967a263dbba116c5d

Signed-off-by: mickhuang <mickhuang@ami.com>
---
 drivers/gpio/gpio-aspeed-sgpio.c | 66 +++++++++-----------------------
 1 file changed, 18 insertions(+), 48 deletions(-)

diff --git a/drivers/gpio/gpio-aspeed-sgpio.c b/drivers/gpio/gpio-aspeed-sgpio.c
index 450d6eb8713d..bf47a2489935 100644
--- a/drivers/gpio/gpio-aspeed-sgpio.c
+++ b/drivers/gpio/gpio-aspeed-sgpio.c
@@ -23,7 +23,6 @@
 #define SGPIO_G7_CTRL_REG_BASE 0x80
 #define SGPIO_G7_CTRL_REG_OFFSET(x) (SGPIO_G7_CTRL_REG_BASE + (x) * 0x4)
 #define SGPIO_G7_OUT_DATA BIT(0)
-#define SGPIO_G7_PARALLEL_OUT_DATA BIT(1)
 #define SGPIO_G7_IRQ_EN BIT(2)
 #define SGPIO_G7_IRQ_TYPE0 BIT(3)
 #define SGPIO_G7_IRQ_TYPE1 BIT(4)
@@ -35,14 +34,6 @@
 #define SGPIO_G7_IRQ_STS BIT(12)
 #define SGPIO_G7_IN_DATA BIT(13)
 #define SGPIO_G7_PARALLEL_IN_DATA BIT(14)
-#define SGPIO_G7_SERIAL_OUT_SEL GENMASK(17, 16)
-#define SGPIO_G7_PARALLEL_OUT_SEL GENMASK(19, 18)
-#define SELECT_FROM_CSR 0
-#define SELECT_FROM_PARALLEL_IN 1
-#define SELECT_FROM_SERIAL_IN 1
-
-#define BMC_CONTROL_START_INDEX 128
-#define BMC_CONTROL_END_INDEX 143
 
 static inline u32 field_get(u32 _mask, u32 _val)
 {
@@ -83,7 +74,6 @@ struct aspeed_sgpio {
 	void __iomem *base;
 	int irq;
 	int version;
-	const struct aspeed_sgpio_pdata *pdata;
 };
 
 struct aspeed_sgpio_bank {
@@ -215,14 +205,6 @@ static bool aspeed_sgpio_is_input(unsigned int offset)
 	return !(offset % 2);
 }
 
-static bool aspeed_sgpios_ctrl_by_csr(unsigned int offset)
-{
-	if (offset >= BMC_CONTROL_START_INDEX &&
-	    offset <= BMC_CONTROL_END_INDEX)
-		return true;
-	return false;
-}
-
 static int aspeed_sgpio_get(struct gpio_chip *gc, unsigned int offset)
 {
 	struct aspeed_sgpio *gpio = gpiochip_get_data(gc);
@@ -235,13 +217,8 @@ static int aspeed_sgpio_get(struct gpio_chip *gc, unsigned int offset)
 	raw_spin_lock_irqsave(&gpio->lock, flags);
 
 	if (gpio->version == 7) {
-		if (gpio->pdata->slave && !aspeed_sgpios_ctrl_by_csr(offset))
-			reg = aspeed_sgpio_is_input(offset) ?
-				      SGPIO_G7_PARALLEL_IN_DATA :
-				      SGPIO_G7_PARALLEL_OUT_DATA;
-		else
-			reg = aspeed_sgpio_is_input(offset) ? SGPIO_G7_IN_DATA :
-							      SGPIO_G7_OUT_DATA;
+		reg = aspeed_sgpio_is_input(offset) ? SGPIO_G7_IN_DATA :
+					      SGPIO_G7_OUT_DATA;
 		rc = !!(field_get(reg, ioread32(addr)));
 	} else {
 		bank = to_bank(offset);
@@ -285,27 +262,19 @@ static int sgpio_g7_set_value(struct gpio_chip *gc, unsigned int offset,
 {
 	struct aspeed_sgpio *gpio = gpiochip_get_data(gc);
 	void __iomem *addr = gpio->base + SGPIO_G7_CTRL_REG_OFFSET(offset >> 1);
-	u32 reg = 0, out_data;
+	u32 reg = 0;
 
 	if (aspeed_sgpio_is_input(offset))
 		return -EINVAL;
 
-	if (gpio->pdata->slave && !aspeed_sgpios_ctrl_by_csr(offset)) {
-		// Ensure the parallel out value control by the software.
-		ast_write_bits(addr, SGPIO_G7_PARALLEL_OUT_SEL,
-			       SELECT_FROM_CSR);
-		out_data = SGPIO_G7_PARALLEL_OUT_DATA;
-	} else {
-		// Ensure the serial out value control by the software.
-		ast_write_bits(addr, SGPIO_G7_SERIAL_OUT_SEL, SELECT_FROM_CSR);
-		out_data = SGPIO_G7_OUT_DATA;
-	}
+	// Ensure the serial out value control by the software.
+	ast_clr_bits(addr, SGPIO_G7_HW_BYPASS_EN | SGPIO_G7_HW_IN_SEL);
 	reg = ioread32(addr);
 
 	if (val)
-		reg |= out_data;
+		reg |= SGPIO_G7_OUT_DATA;
 	else
-		reg &= ~out_data;
+		reg &= ~SGPIO_G7_OUT_DATA;
 
 	iowrite32(reg, addr);
 
@@ -851,6 +820,7 @@ MODULE_DEVICE_TABLE(of, aspeed_sgpio_of_table);
 static int aspeed_sgpio_probe(struct platform_device *pdev)
 {
 	u32 nr_gpios, sgpio_freq, sgpio_clk_div, gpio_cnt_regval, pin_mask;
+	const struct aspeed_sgpio_pdata *pdata;
 	struct aspeed_sgpio *gpio;
 	unsigned long apb_freq;
 	void __iomem *addr;
@@ -866,12 +836,12 @@ static int aspeed_sgpio_probe(struct platform_device *pdev)
 
 	gpio->dev = &pdev->dev;
 
-	gpio->pdata = device_get_match_data(&pdev->dev);
-	if (!gpio->pdata)
+	pdata = device_get_match_data(&pdev->dev);
+	if (!pdata)
 		return -EINVAL;
 
-	pin_mask = gpio->pdata->pin_mask;
-	gpio->version = gpio->pdata->version;
+	pin_mask = pdata->pin_mask;
+	gpio->version = pdata->version;
 
 	rc = device_property_read_u32(&pdev->dev, "ngpios", &nr_gpios);
 	if (rc < 0) {
@@ -883,14 +853,14 @@ static int aspeed_sgpio_probe(struct platform_device *pdev)
 		return -EINVAL;
 	}
 
-	if (gpio->version == 7 && !gpio->pdata->slave)
+	if (gpio->version == 7 && !pdata->slave)
 		for (i = 0; i < nr_gpios; i++) {
 			addr = gpio->base + SGPIO_G7_CTRL_REG_OFFSET(i);
-			ast_write_bits(addr, SGPIO_G7_SERIAL_OUT_SEL,
-				       SELECT_FROM_CSR);
+			ast_clr_bits(addr, SGPIO_G7_HW_BYPASS_EN |
+						   SGPIO_G7_HW_IN_SEL);
 		}
 
-	if (!gpio->pdata->slave) {
+	if (!pdata->slave) {
 		rc = device_property_read_u32(&pdev->dev, "bus-frequency", &sgpio_freq);
 		if (rc < 0) {
 			dev_err(&pdev->dev, "Could not read bus-frequency property\n");
@@ -925,9 +895,9 @@ static int aspeed_sgpio_probe(struct platform_device *pdev)
 		gpio_cnt_regval = ((nr_gpios / 8) << ASPEED_SGPIO_PINS_SHIFT) & pin_mask;
 		iowrite32(FIELD_PREP(ASPEED_SGPIO_CLK_DIV_MASK, sgpio_clk_div) |
 				gpio_cnt_regval | ASPEED_SGPIO_ENABLE,
-			gpio->base + gpio->pdata->ctrl_reg);
+			gpio->base + pdata->ctrl_reg);
 	} else {
-		iowrite32(ASPEED_SGPIO_ENABLE, gpio->base + gpio->pdata->ctrl_reg);
+		iowrite32(ASPEED_SGPIO_ENABLE, gpio->base + pdata->ctrl_reg);
 	}
 
 	raw_spin_lock_init(&gpio->lock);
-- 
2.34.1

